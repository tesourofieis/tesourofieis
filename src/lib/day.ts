import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  PATTERN_ADVENT,
  PATTERN_EASTER,
  PATTERN_EPIPHANY,
  PATTERN_LENT,
  PATTERN_POST_EPIPHANY_SUNDAY,
  PATTERN_PRE_LENTEN,
  TABLE_OF_PRECEDENCE,
  TEMPORA_RANK_MAP,
  TITLES,
  TYPE_TEMPORA,
} from "./constants.ts";

// """ Class used to keep `Observance` objects for particular days of Missal.
//
// It contains three lists: `tempora`, `celebration` and `commemoration`.
// On Missal's creation the lists are filled in so that `tempora` always contains `Observance` representing
// given variable day, `celebration` contains an `Observance`s to be celebrated in this day and
// `commemoration` contains zero or more `Observance`s that should be commemorated with the main celebration.
// """
class Day {
  date: string;
  tempora: Observance[] = [];
  celebration: Observance[] = [];
  commemoration: Observance[] = [];

  constructor(date: string) {
    this.date = date;
  }

  get all(): Observance[] {
    return this.tempora.concat(this.celebration).concat(this.commemoration);
  }

  getTemporaName() {
    return this.tempora[0]?.title;
  }

  serialize() {
    const serialized = {};
    const containers = ["tempora", "celebration", "commemoration"];

    for (const container of containers) {
      serialized[container] = this[container].map((item: Day) =>
        item.serialize(),
      );
    }

    return serialized;
  }
}

class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  link: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, Number.parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.link = `missal/${this.getTempora(this.name)}/${this.name}`;
    this.title = TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      if (name.replace(/^.*-(\d+).*$/, "$1") === "10-Dur") {
        this.weekday = 0;
      }
      this.weekday = Number.parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10);
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  private getTempora(name: string): string {
    if (this.flexibility === "tempora") {
      if (name.includes("nat")) {
        return "natal";
      }
      if (name.includes("adv")) {
        return "advento";
      }
      if (name.includes("pasc")) {
        return "pascoa";
      }
      if (name.includes("epi")) {
        return "epifania";
      }
      if (name.includes("quadp")) {
        return "pre-quaresma";
      }
      if (name.includes("quad")) {
        return "quaresma";
      }
      if (name.includes("pent")) {
        return "pentecostes";
      }
    }

    if (this.flexibility === "santos") {
      console.log(`Matched Santos: ${name}`);
      return "santos";
    }

    if (this.flexibility === "commune") {
      return "comum";
    }

    if (this.flexibility === "votivas") {
      return "votivas";
    }

    console.log(`Checking name: ${name}, flexibility: ${this.flexibility}`);
    throw Error(`No directory for ${name}, flexibility , ${this.flexibility}`);
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      link: this.link,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}

export { Day, Observance };
