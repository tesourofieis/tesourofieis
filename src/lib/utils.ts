import { CUSTOM_PREFACES, PROPERS_DIR } from "./constants.ts";
import path from "path";
import fs from "fs";
import { Observance } from "./observance.ts";
import { Proper } from "./proper.ts";
import { Day } from "./day.ts";
import type { Calendar } from "./calendar.ts";
import type { UTCDate } from "@date-fns/utc";
import { format } from "date-fns";

type Pattern = string | RegExp;

function match(
  observances: string | Observance | (string | Observance)[],
  patterns: string | Pattern | (string | Pattern)[],
): Observance | undefined {
  if (!Array.isArray(observances)) {
    observances = [observances];
  }

  if (!Array.isArray(patterns)) {
    patterns = [patterns];
  }

  for (const observance of observances) {
    const observanceId =
      typeof observance === "string" ? observance : observance.id;

    for (const pattern of patterns) {
      if (
        typeof pattern === "string" &&
        new RegExp(pattern).test(observanceId)
      ) {
        return observance as Observance;
      }
      if (pattern instanceof RegExp && pattern.test(observanceId)) {
        return observance as Observance;
      }
    }
  }
}

function yyyyMMDD(date: Date | UTCDate) {
  return format(date, "yyyy-MM-dd");
}

function getCustomPreface(
  celebration: Observance,
  tempora?: Observance,
): string | null {
  for (const { id, preface } of CUSTOM_PREFACES) {
    if (
      new RegExp(id).test(celebration.id) ||
      (tempora && celebration.rank > 1 && new RegExp(id).test(tempora.id))
    ) {
      return preface;
    }
  }
  return null;
}

function getPregeneratedProper(lang: string, properId: string) {
  const filePath = path.join(PROPERS_DIR, lang, `${properId}.json`);

  if (fs.existsSync(filePath)) {
    const fileContent = fs.readFileSync(filePath, "utf-8");
    return JSON.parse(fileContent);
  }

  return null;
}

export interface ProperDay {
  info: {
    id: string;
    title: string;
    tempora: string;
    rank: number;
    colors: string[];
    date: string;
    day: Day;
  };
  sections: Section[];
}

interface Section {
  id: string;
  body: {
    latin: string[];
    vernacular: string[];
  };
}

function format_propers(propers: Proper[][], day: Day): ProperDay[] {
  return propers.map((proper) => {
    const [properVernacular, properLatin] = proper;

    const title = properVernacular.title;
    const temporaName = day.getTemporaName();

    const info = {
      id: properVernacular.id,
      title: title || "Feria",
      tempora: temporaName !== title ? temporaName : null,
      rank: properVernacular.rank,
      colors: properVernacular.colors,
      date: day.date,
      day: day,
    };

    return {
      info: info,
      sections: format_proper_section(properLatin, properVernacular),
    };
  });
}

function format_proper_section(
  propers_latin: Proper,
  propers_vernacular: Proper,
) {
  const pl = [];
  const latinPropers = propers_latin.serialize();
  const vernacularPropers = propers_vernacular.serialize();

  for (const vernacular of vernacularPropers) {
    const latinProp = latinPropers.find(
      (latin) => latin?.id === vernacular?.id,
    );
    const latinBody = latinProp?.body;
    const vernacularBody = vernacular?.body;

    pl.push({
      id: vernacular?.id,
      body: {
        latin: latinBody,
        vernacular: vernacularBody,
      },
    });
  }

  return pl;
}

function printAll(calendar: Calendar) {
  return calendar.serialize();
}

export {
  match,
  getCustomPreface,
  getPregeneratedProper,
  yyyyMMDD,
  format_propers,
  printAll,
};
