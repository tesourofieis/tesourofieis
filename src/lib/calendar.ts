import { Day, Observance } from "./day.ts";
import { yyyyMMDD } from "./utils.ts";

import {
  ADVENT,
  EMBER_DAYS_SEPTEMBER,
  FROM_PRE_LENT_TO_POST_PENTECOST,
  LOCAL,
  NATIVITY_OCTAVE_SUNDAY,
  OUTRO,
  POST_EPIPHANY,
  SANCTI,
  SANCTI_10_DUr,
  TEMPORA_NAT2_0,
  WEEK_24_AFTER_PENTECOST,
} from "./constants.ts";
import { rules } from "./rules.ts";

import { UTCDate } from "@date-fns/utc";
import {
  addDays,
  format,
  getDate,
  getDay,
  getMonth,
  getYear,
  isSameDay,
  isSunday,
  nextSunday,
  nextWednesday,
  previousSaturday,
  previousSunday,
  subDays,
} from "date-fns";

class Calendar {
  private container: Map<string, Day>;
  public year: number;

  constructor(year: number) {
    this.year = year;
    this.container = new Map();
    this.buildEmptyCalendar(year);
    this.create();
  }

  private buildEmptyCalendar(year: number): void {
    let date = new UTCDate(year, 0, 1);

    while (date.getFullYear() === year) {
      this.container.set(yyyyMMDD(date), new Day(yyyyMMDD(date)));
      date = addDays(date, 1);
    }
  }

  private create() {
    this.fillInTemporaDays();
    this.fillInSanctiDays();
    this.resolveConcurrency();
  }

  private fillInTemporaDays() {
    // """
    // Days depending on variable date, such as Easter or Advent
    // """
    // # Inserting blocks
    this.insertBlock(this.calcHolyFamily(this.year), POST_EPIPHANY);
    this.insertBlock(
      this.calcSeptuagesima(this.year),
      FROM_PRE_LENT_TO_POST_PENTECOST,
    );
    this.insertBlock(
      this.calcSaturdayBefore24SundayAfterPentecost(this.year),
      POST_EPIPHANY,
      true,
      false,
      this.calcFirstAdventSunday(this.year),
    );
    this.insertBlock(
      this.calc24SundayAfterPentecost(this.year),
      WEEK_24_AFTER_PENTECOST,
    );
    this.insertBlock(
      this.calcFirstAdventSunday(this.year),
      ADVENT,
      false,
      false,
      new UTCDate(this.year, 11, 23),
    );
    this.insertBlock(
      this.calcEmberWednesdaySeptember(this.year),
      EMBER_DAYS_SEPTEMBER,
    );

    // # Inserting single days
    const holyNameDate = this.calcHolyName(this.year);

    this.container.get(yyyyMMDD(holyNameDate)).celebration = [
      new Observance(TEMPORA_NAT2_0, yyyyMMDD(holyNameDate)),
    ];

    const christKingDate = this.calcChristKing(this.year);
    if (this.container.get(yyyyMMDD(christKingDate))) {
      this.container.get(yyyyMMDD(christKingDate)).celebration = [
        new Observance(SANCTI_10_DUr, yyyyMMDD(christKingDate)),
      ];
    }

    const christmasOctaveSunday = this.calcSundayChristmasOctave(this.year);
    if (christmasOctaveSunday) {
      this.insertBlock(christmasOctaveSunday, NATIVITY_OCTAVE_SUNDAY);
    }
  }

  private insertBlock(
    date: Date,
    block: string[],
    reverse = false,
    overwrite = true,
    stopDate?: Date,
  ) {
    if (reverse) {
      // TODO: use toReversed in order not to mutate the original
      // or slice
      block = block.slice().reverse();
    }

    for (const [ii, observanceIds] of block.entries()) {
      const currentDate = addDays(new UTCDate(date), reverse ? -ii : ii);

      if (!observanceIds) {
        continue;
      }

      const dateKey = yyyyMMDD(currentDate);

      if (this.container.get(dateKey)?.celebration.length && !overwrite) {
        break;
      }

      if (isSameDay(stopDate, subDays(currentDate, 1))) {
        break;
      }

      if (this.container.get(dateKey)) {
        this.container.get(dateKey).tempora = [
          new Observance(observanceIds, dateKey),
        ];
        this.container.get(dateKey).celebration = [
          ...this.container.get(dateKey).tempora,
        ];
      }
    }
  }

  private fillInSanctiDays() {
    for (const [date] of this.container.entries()) {
      const dateId = format(date, "MM-dd");
      const days = SANCTI.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      this.container.get(date)?.celebration.push(...days);

      const local = LOCAL.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (local.length) {
        this.container.get(date)?.local.push(...local);
      }

      const outro = OUTRO.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (outro.length) {
        this.container.get(date)?.outro.push(...outro);
      }
    }
  }

  // Apply `kalendar.rules.*` to the initially instantiated Calendar to fix the situations
  // where more than one Observance falls in the same day.
  private resolveConcurrency() {
    const shiftedAll: { [date: string]: Observance[] } = {};

    for (const [date] of this.container.entries()) {
      const [celebration, commemoration, shifted] = this.applyRules(
        date,
        shiftedAll[date] || [],
      );

      this.container.get(date).celebration = celebration as Observance[];
      this.container.get(date).commemoration = commemoration as Observance[];

      if (shifted) {
        // @ts-ignore
        for (const [k, v] of shifted) {
          if (!shiftedAll[k]) {
            shiftedAll[k] = [];
          }
          shiftedAll[k].push(...(Array.isArray(v) ? v : [v]));
        }
      }
    }
  }

  private applyRules(date: string, shifted: Observance[]) {
    for (const rule of rules) {
      const results = rule(
        this,
        date,
        this.container.get(date)?.tempora,
        this.container.get(date).celebration.concat(shifted),
      );

      if (results) {
        return results;
      }
    }

    return [this.container.get(date).celebration, [], []];
  }

  private calcEasterSunday(year: number): Date {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const month = Math.floor((h + l - 7 * m + 114) / 31);
    const day = ((h + l - 7 * m + 114) % 31) + 1;
    const easter = new UTCDate(year, month - 1, day);

    return easter;
  }

  private calcHolyFamily(year: number): Date {
    // Feast of the Holy Family - First Sunday after Epiphany (06 January).
    const d = new UTCDate(year, 0, 6); // Months are zero-indexed in JavaScript

    return nextSunday(d);
  }

  private calcSeptuagesima(year: number): Date {
    // """ Septuagesima Sunday.
    //
    // Beginning of the pre-Lenten season (Shrovetide).
    // It's 63 days before Ressurection, ninth Sunday before Easter, the third before Ash Wednesday.
    // First day of the Ressurection Sunday - related block.
    // """
    const easter = this.calcEasterSunday(year);
    return subDays(easter, 63);
  }

  private calcFirstAdventSunday(year: number): Date {
    // """
    // First Sunday of Advent - November 27 if it's Sunday, otherwise closest Sunday.
    // """
    const christmasDay = new UTCDate(year, 11, 25);
    const weekday = getDay(christmasDay);

    switch (weekday) {
      case 0: // Sunday
        return new UTCDate(getYear(christmasDay), 10, 27);
      case 1: // Monday
        return new UTCDate(getYear(christmasDay), 11, 3);
      case 2: // Tuesday
        return new UTCDate(getYear(christmasDay), 11, 2);
      case 3: // Wednesday
        return new UTCDate(getYear(christmasDay), 11, 1);
      case 4: // Thursday
        return new UTCDate(getYear(christmasDay), 10, 30);
      case 5: // Friday
        return new UTCDate(getYear(christmasDay), 10, 29);
      default: // Saturday
        return new UTCDate(getYear(christmasDay), 10, 28);
    }
  }

  private calc24SundayAfterPentecost(year: number): Date {
    // """ 24th Sunday after Pentecost.
    //
    // Last Sunday before First Sunday of Advent.
    // It will be always TEMPORA_PENT24_0, which will be placed either:
    // * instead of TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 23)
    // * directly after a week starting with TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 24)
    // * directly after a week starting with TEMPORA_EPI6_0 (moved from post-epiphania period)
    //   if the number of TEMPORA_PENT*_0 Sundays in given year > 24)
    // """
    return previousSunday(new UTCDate(this.calcFirstAdventSunday(year)));
  }

  private calcSaturdayBefore24SundayAfterPentecost(year: number): Date {
    // """ Last Saturday before 24th Sunday after Pentecost.
    //
    // This is the end of potentially "empty" period that might appear
    // between 23rd and 24th Sunday after Pentecost if Easter is early.
    // In such case one or more Sundays after Epiphany (TEMPORA_EPI*_0) are moved here to "fill the gap"

    return previousSaturday(new UTCDate(this.calc24SundayAfterPentecost(year)));
  }

  private calcEmberWednesdaySeptember(year: number): Date {
    // """ Wednesday of the Ember Days of September.
    //
    // Ember Wednesday in September is a Wednesday after third Sunday
    // of September according to John XXIII's motu proprio
    // "Rubricarum instructum" of June 25 1960.
    // """
    let d = new UTCDate(year, 8, 1);
    while (getMonth(d) === 8) {
      if (isSunday(d) && getDate(d) >= 15 && getDate(d) <= 21) {
        break;
      }
      d = addDays(d, 1);
    }
    return nextWednesday(d);
  }

  private calcHolyName(year: number) {
    // """ The Feast of the Holy Name of Jesus.
    //
    // Kept on the First Sunday of the year; but if this Sunday falls on
    // 1st, 6th or 7th January, the feast is kept on 2nd January.
    // """
    let d = new UTCDate(year, 0, 1);

    while (getDate(d) <= 7) {
      if (isSunday(d)) {
        if (getDate(d) === 1 || getDate(d) === 6 || getDate(d) === 7) {
          return new UTCDate(year, 0, 2);
        }

        return d;
      }

      d = addDays(d, 1);
    }

    console.error("Unexpected condition: Holy Name date not found.");
  }

  private calcChristKing(year: number): Date {
    // The Feast of Christ the King, last Sunday of October.
    const d = new UTCDate(year, 10, 1);

    return previousSunday(d);
  }

  private calcSundayChristmasOctave(year: number): Date | null {
    // """
    // Sunday within the Octave of Christmas, falls between Dec 26 and Dec 31
    // """
    let d = new UTCDate(year, 11, 27); // December 27
    while (d.getFullYear() === year) {
      if (isSunday(d)) {
        return d;
      }
      d = addDays(d, 1);
    }
    return null;
  }

  get(date: string) {
    return this.container.get(date);
  }

  findDay(observanceId: string) {
    for (const [date_, day] of this.container) {
      if (day.all.some((observance) => observance.id === observanceId)) {
        return [date_, day];
      }
    }
  }

  private items() {
    return this.container.entries();
  }

  serialize(): { [date: string]: Day } {
    const serialized = {};
    for (const [date_, day] of this.items()) {
      serialized[date_] = day.serialize();
    }
    return serialized;
  }
}

export { Calendar };
