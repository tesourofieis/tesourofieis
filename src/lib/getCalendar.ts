import { Calendar } from "./src/calendar.ts";

function getCalendar(year: number) {
  const calendar = new Calendar(year);

  return calendar;
}

function getCalendarDay(year: number, date: string) {
  const calendar = new Calendar(year).get(date);

  return calendar;
}

export { getCalendar, getCalendarDay };
