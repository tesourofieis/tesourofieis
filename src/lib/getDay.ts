import { getYear } from "date-fns";
import { Calendar } from "./calendar.ts";
import { formatPropers } from "./utils.ts";

function getDay(date: string) {
  const calendar = new Calendar(getYear(date));

  const day = calendar.get(date);

  const proper = day?.getProper(calendar);

  const format = formatPropers(proper, day);

  return format;
}

export { getDay };
