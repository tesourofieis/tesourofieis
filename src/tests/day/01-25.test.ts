import { expect, test } from "vitest";
import { getDay } from "../../lib/getDay";

test("2024-02-25", () => {
  const proper = getDay("2024-02-25");

  expect(proper).toMatchSnapshot();
});
