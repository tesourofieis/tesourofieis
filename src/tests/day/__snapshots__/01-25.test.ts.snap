// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2024-02-25 1`] = `
[
  {
    "info": {
      "colors": [
        "v",
      ],
      "date": "2024-02-25",
      "day": Day {
        "celebration": [
          Observance {
            "colors": [
              "v",
            ],
            "date": "2024-02-25",
            "flexibility": "tempora",
            "id": "tempora:Quad2-0:1:v",
            "name": "Quad2-0",
            "priority": 19,
            "rank": 1,
            "title": "2º Domingo da Quaresma",
            "weekday": 0,
          },
        ],
        "commemoration": [],
        "date": "2024-02-25",
        "tempora": [
          Observance {
            "colors": [
              "v",
            ],
            "date": "2024-02-25",
            "flexibility": "tempora",
            "id": "tempora:Quad2-0:1:v",
            "name": "Quad2-0",
            "priority": 19,
            "rank": 1,
            "title": "2º Domingo da Quaresma",
            "weekday": 0,
          },
        ],
      },
      "id": "tempora:Quad2-0:1:v",
      "rank": 1,
      "tempora": null,
      "title": "2º Domingo da Quaresma",
    },
    "sections": [
      {
        "body": {
          "latin": [
            "Ps 24:6; 24:3; 24:22",
            "Reminíscere miseratiónum tuarum, Dómine, et misericórdiæ tuæ, quæ a sǽculo sunt: ne umquam dominéntur nobis inimíci nostri: líbera nos, Deus Israël, ex ómnibus angústiis nostris.",
            "Ps 24:1-2",
            "Ad te, Dómine, levávi ánimam meam: Deus meus, in te confído, non erubéscam.",
            "Glória Patri…",
            "Reminíscere miseratiónum tuarum, Dómine, et misericórdiæ tuæ, quæ a sǽculo sunt: ne umquam dominéntur nobis inimíci nostri: líbera nos, Deus Israël, ex ómnibus angústiis nostris.",
            "",
          ],
          "vernacular": [
            "Sl 24:6, 3, 22.",
            "Lembrai-Vos, Senhor, de que a vossa bondade e misericórdia são eternas! Que os nossos inimigos nunca triunfem de nós. Ó Deus de Israel, livrai-nos de todas nossas angústias.",
            "Sl 24:1-2",
            "A Vós, Senhor, elevei a minha alma: meu Deus, confio em Vós; não ficarei confundido.",
            "Glória ao Pai.",
            "Lembrai-Vos, Senhor, de que a vossa bondade e misericórdia são eternas! Que os nossos inimigos nunca triunfem de nós. Ó Deus de Israel, livrai-nos de todas nossas angústias.",
            "",
          ],
        },
        "id": "Introitus",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Deus, qui cónspicis omni nos virtúte destítui: intérius exteriúsque custódi; ut ab ómnibus adversitátibus muniámur in córpore, et a pravis cogitatiónibus mundémur in mente.",
            "Per Dominum…",
            "",
          ],
          "vernacular": [
            "Ó Deus, que conheceis como somos destituídos de toda a virtude, guardai-nos interior e exteriormente, a fim de que o nosso corpo seja preservado de todas as adversidades e a nossa alma purificada de todos os maus pensamentos.",
            "Por nosso Senhor...",
            "",
          ],
        },
        "id": "Oratio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Léctio Epístolæ beáti Pauli Apóstoli ad Thessalonicénses",
            "1 Thess 4:1-7.",
            "Fratres: Rogámus vos et obsecrámus in Dómino Jesu: ut, quemádmodum accepístis a nobis, quómodo opórteat vos ambuláre et placére Deo, sic et ambulétis, ut abundétis magis. Scitis enim, quæ præcépta déderim vobis Per Dominum Jesum. Hæc est enim volúntas Dei, sanctificátio vestra: ut abstineátis vos a fornicatióne, ut sciat unusquísque vestrum vas suum possidére in sanctificatióne et honóre; non in passióne desidérii, sicut et gentes, quæ ignórant Deum: et ne quis supergrediátur neque circumvéniat in negótio fratrem suum: quóniam vindex est Dóminus de his ómnibus, sicut prædíximus vobis et testificáti sumus. Non enim vocávit nos Deus in immundítiam, sed in sanctificatiónem: in Christo Jesu, Dómino nostro.",
            "",
          ],
          "vernacular": [
            "Lição da Ep.ª do B. Ap.º Paulo aos Tessalonicenses.",
            "1 Ts 4:1-7",
            "Meus irmãos: Vos pedimos e exortamos, em nome do Senhor Jesus, que, havendo aprendido de nós como deveis conduzir-vos para agradar a Deus, tenhais uma conduta de modo a aperfeiçoar-vos cada vez mais. Com efeito, conheceis os preceitos que vos dei da parte do Senhor Jesus. O que Ele quer é a vossa santificação: que vos abstenhais da impureza carnal, para que cada um saiba guardar o vaso do seu corpo em santidade e honestidade, e não segundo os apetites das paixões, como os pagãos, que desconhecem Deus; e que ninguém engane o seu irmão, nem o prejudique, pois, o Senhor é vingador destas coisas, como já provámos e testemunhámos. Porquanto Deus vos não chamou para os gozos da carne, mas para a santificação em nosso Senhor Jesus Cristo.",
            "",
          ],
        },
        "id": "Lectio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 24:17-18",
            "Tribulatiónes cordis mei dilatátæ sunt: de necessitátibus meis éripe me, Dómine,",
            "℣. Vide humilitátem meam et labórem meum: et dimítte ómnia peccáta mea.",
            "",
            "Tractus",
            "Ps 105:1-4",
            "Confitémini Dómino, quóniam bonus: quóniam in sǽculum misericórdia ejus.",
            "℣. Quis loquétur poténtias Dómini: audítas fáciet omnes laudes ejus?",
            "℣. Beáti, qui custódiunt judícium et fáciunt justítiam in omni témpore.",
            "℣. Meménto nostri, Dómine, in beneplácito pópuli tui: vísita nos in salutári tuo.",
            "",
          ],
          "vernacular": [
            "Sl 24:17-18",
            "As tribulações do meu coração cresceram. Ó Senhor, livrai-me das minhas misérias.",
            "℣. Vede a minha humilhação e fadiga e perdoai os meus pecados.",
            "",
            "Tractus",
            "Sl 105:1-4",
            "Louvai o Senhor, pois Ele é bom: a sua misericórdia é eterna.",
            "℣. Quem será capaz de narrar as maravilhas da omnipotência do Senhor e apregoar os seus louvores?",
            "℣. Bem-aventurados aqueles que procedem com equidade e justiça em todas as ocasiões.",
            "℣. Pela vossa bondade para com vosso povo, Senhor, lembrai-Vos de nós: visitai-nos para alcançarmos a salvação.",
            "",
          ],
        },
        "id": "Graduale",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Sequéntia ☩ sancti Evangélii secúndum Matthǽum",
            "Matt 17:1-9",
            "In illo témpore: Assúmpsit Jesus Petrum, et Jacóbum, et Joánnem fratrem eius, et duxit illos in montem excélsum seórsum: et transfigurátus est ante eos. Et resplénduit fácies ejus sicut sol: vestiménta autem ejus facta sunt alba sicut nix. Et ecce, apparuérunt illis Móyses et Elías cum eo loquéntes. Respóndens autem Petrus, dixit ad Jesum: Dómine, bonum est nos hic esse: si vis, faciámus hic tria tabernácula, tibi unum, Móysi unum et Elíæ unum. Adhuc eo loquénte, ecce, nubes lúcida obumbrávit eos. Et ecce vox de nube, dicens: Hic est Fílius meus diléctus, in quo mihi bene complácui: ipsum audíte. Et audiéntes discípuli, cecidérunt in fáciem suam, et timuérunt valde. Et accéssit Jesus, et tétigit eos, dixítque eis: Súrgite, et nolíte timére. Levántes autem óculos suos, néminem vidérunt nisi solum Jesum. Et descendéntibus illis de monte, præcépit eis Jesus, dicens: Némini dixéritis visiónem, donec Fílius hóminis a mórtuis resúrgat.",
            "",
          ],
          "vernacular": [
            "Continuação ☩ do santo Evangelho segundo S. Mateus.",
            "Mt. 17:1-9",
            "Naquele tempo, Jesus levou consigo Pedro, Tiago e João e conduziu-os a um monte alto e separado, transfigurando-se ante eles: seu rosto resplandecia, como o sol, e os seus vestidos tornaram-se brancos, como a neve! E Moisés e Elias apareceram, conversando com Jesus. Então, Pedro disse a Jesus: «Senhor, é tão bom estar aqui!... Se quereis, façamos aqui três tendas: uma para Vós, outra para Moisés e outra para Elias!». Ainda ele falava, eis que uma nuvem brilhante os envolveu, saindo do seio dela uma voz, que dizia: «Este é o meu Filho muito amado, em quem pus as minhas complacências; ouvi-O». Havendo escutado a voz, os discípulos caíram com o rosto no chão e ficaram atemorizados. Mas Jesus tocou-os e disse-lhes: «Levantai-vos; não vos amedronteis». Então, erguendo os olhos, já nada viram senão só Jesus. Desceram do monte, dando-lhes Jesus esta ordem: «Não conteis a ninguém esta visão até que o Filho do homem ressuscite dos mortos».",
            "",
          ],
        },
        "id": "Evangelium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 118:47; 118:48",
            "Meditábor in mandátis tuis, quæ diléxi valde: et levábo manus meas ad mandáta tua, quæ diléxi.",
            "",
          ],
          "vernacular": [
            "Sl 118:47-48",
            "Meditarei nos vossos Mandamentos, que muito amo: e levantarei as minhas mãos para cumprir esses Mandamentos, que, repito, muito amo.",
            "",
          ],
        },
        "id": "Offertorium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Sacrifíciis præséntibus, Dómine, quǽsumus, inténde placátus: ut et devotióni nostræ profíciant et salúti.",
            "Per Dominum…",
            "",
          ],
          "vernacular": [
            "Senhor, Vos suplicamos, dignai-Vos olhar benigno para o presente sacrifício, a fim de que sirva de proveito à nossa piedade e à nossa salvação.",
            "Por nosso Senhor...",
            "",
          ],
        },
        "id": "Secreta",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "de Quadragesima",
            "Vere dignum et justum est, æquum et salutáre, nos tibi semper et ubíque grátias ágere: Dómine, sancte Pater, omnípotens ætérne Deus: Qui corporáli ieiúnio vítia cómprimis, mentem élevas, virtútem largíris et prǽmia: per Christum Dóminum nostrum. Per quem majestátem tuam laudant Angeli, adórant Dominatiónes, tremunt Potestátes. Cæli cælorúmque Virtútes, ac beáta Séraphim, sócia exsultatióne concélebrant. Cum quibus et nostras voces, ut admítti iúbeas, deprecámur, súpplici confessióne dicéntes:",
            "",
          ],
          "vernacular": [
            "da Quaresma",
            "Verdadeiramente é digno e justo, racional e salutar dar-Vos graças em todos os lugares e sempre, Senhor santo, Pai omnipotente, eterno Deus, pois Vós, pelo jejum natural, reprimis os vícios, elevais o espírito e concedeis-nos a virtude e o prémio por Cristo, nosso Senhor: pelo qual os Anjos louvam a vossa majestade, as Dominações a adoram e as Potestades trémulas a reverenciam; e os Céus, as Virtudes dos Céus e os bem-aventurados Serafins se associam em comum louvor. Dignai-Vos permitir, Senhor, que as nossas vozes suplicantes se unam às deles, dizendo:",
            "",
          ],
        },
        "id": "Prefatio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 5:2-4",
            "Intéllege clamórem meum: inténde voci oratiónis meæ, Rex meus et Deus meus: quóniam ad te orábo, Dómine.",
            "",
          ],
          "vernacular": [
            "Sl 5:2-4",
            "Ouvi o meu clamor; ouvi a minha oração, ó meu Rei, ó meu Deus; pois orei a Vós.",
            "",
          ],
        },
        "id": "Communio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Súpplices te rogámus, omnípotens Deus: ut quos tuis réficis sacraméntis, tibi étiam plácitis móribus dignánter deservíre concédas.",
            "Per Dominum…",
            "",
          ],
          "vernacular": [
            "Ó Deus omnipotente, Vos suplicamos instantemente que aqueles a quem sustentais com vossos sacramentos tenham uma conduta de vida que Vos seja agradável.",
            "Por nosso Senhor...",
            "",
          ],
        },
        "id": "Postcommunio",
        "subSections": [],
      },
    ],
  },
]
`;
